be/src/exec/pipeline/sink/export_sink_operator.cpp

```diff
@@ -0,0 +1,205 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Inc.
+
+#include "exec/pipeline/sink/export_sink_operator.h"
+
+#include "exec/data_sink.h"
+#include "exec/file_builder.h"
+#include "exec/pipeline/sink/sink_io_buffer.h"
+#include "exec/plain_text_builder.h"
+#include "formats/csv/converter.h"
+#include "formats/csv/output_stream.h"
+#include "fs/fs_broker.h"
+#include "runtime/runtime_state.h"
+
+namespace starrocks {
+namespace pipeline {
+
+class ExportSinkIOBuffer final : public SinkIOBuffer {
+public:
+    ExportSinkIOBuffer(const TExportSink& t_export_sink, std::vector<ExprContext*>& output_expr_ctxs,
+                       int32_t num_sinkers)
+            : SinkIOBuffer(num_sinkers), _t_export_sink(t_export_sink), _output_expr_ctxs(output_expr_ctxs) {}
+
+    ~ExportSinkIOBuffer() override = default;
+
+    Status prepare(RuntimeState* state, RuntimeProfile* parent_profile) override;
+
+    void close(RuntimeState* state) override;
+
+private:
+    void _process_chunk(bthread::TaskIterator<const vectorized::ChunkPtr>& iter) override;
+
+    Status _open_file_writer(int timeout_ms);
+
+    Status _gen_file_name(std::string* file_name);
+
+    TExportSink _t_export_sink;
+    const std::vector<ExprContext*> _output_expr_ctxs;
+    std::unique_ptr<FileBuilder> _file_builder;
+};
+
+Status ExportSinkIOBuffer::prepare(RuntimeState* state, RuntimeProfile* parent_profile) {
+    bool expected = false;
+    if (!_is_prepared.compare_exchange_strong(expected, true)) {
+        return Status::OK();
+    }
+    _state = state;
+
+    bthread::ExecutionQueueOptions options;
+    options.executor = SinkIOExecutor::instance();
+    _exec_queue_id = std::make_unique<bthread::ExecutionQueueId<const vectorized::ChunkPtr>>();
+    int ret = bthread::execution_queue_start<const vectorized::ChunkPtr>(_exec_queue_id.get(), &options,
+                                                                         &SinkIOBuffer::execute_io_task, this);
+    if (ret != 0) {
+        return Status::InternalError("start execution queue error");
+    }
+
+    return Status::OK();
+}
+
+void ExportSinkIOBuffer::close(RuntimeState* state) {
+    if (_file_builder != nullptr) {
+        set_io_status(_file_builder->finish());
+        _file_builder.reset();
+    }
+    SinkIOBuffer::close(state);
+}
+
+void ExportSinkIOBuffer::_process_chunk(bthread::TaskIterator<const vectorized::ChunkPtr>& iter) {
+    --_num_pending_chunks;
+    if (_is_finished) {
+        return;
+    }
+
+    if (_is_cancelled && !_is_finished) {
+        close(_state);
+        return;
+    }
+
+    if (_file_builder == nullptr) {
+        int query_timeout = _state->query_options().query_timeout;
+        int timeout_ms = query_timeout > 3600 ? 3600000 : query_timeout * 1000;
+        if (Status status = _open_file_writer(timeout_ms); !status.ok()) {
+            set_io_status(status);
+            close(_state);
```

Comment:
what if query_timeout = 0?

Reply:
> what if query_timeout = 0?

If the user actively sets it to 0, the query will fail immediately